apps/condo/domains/user/integration/sberid/SberIdIdentityIntegration.js

```diff
@@ -0,0 +1,174 @@
+        // set issue token request parameters
+        const request = {
+            grant_type: 'authorization_code',
+            code,
+            redirect_uri: redirectUri,
+            client_id: clientId,
+            client_secret: clientSecret,
+        }
+
+        // send a request
+        const tokenResponse = await axios.create({
+            timeout: axiosTimeout,
+            headers: {
+                'x-ibm-client-id': clientId,
+                'rquid': this.generateRequestId(),
+            },
+            validateStatus: () => true,
+            httpsAgent,
+        }).post(tokenUrl, new URLSearchParams(request))
+
+        // extract required params
+        const {
+            access_token: accessToken,
+            token_type: tokenType,
+            expires_in: expiresIn,
+            scope,
+            id_token: idToken,
+        } = tokenResponse.data
+
+        if (tokenResponse.status !== 200 || isNil(accessToken) || isNil(idToken)) {
+            throw new Error(JSON.stringify(tokenResponse.data))
+        }
+
+        return {
+            accessToken,
+            tokenType,
+            expiresIn,
+            scope,
+            idToken,
+        }
+    }
+
+    async getUserInfo ({ accessToken }) {
+        // send a request
+        const response = await axios.create({
+            timeout: axiosTimeout,
+            headers: {
+                'Authorization': `Bearer ${accessToken}`,
+                'x-ibm-client-id': clientId,
+                'x-introspect-rquid': this.generateRequestId(),
+            },
+            validateStatus: () => true,
+            httpsAgent,
+        }).get(userInfoUrl)
+
+        // extract params that going to be cleaned
+        const {
+            iss,
+            sub,
+            email,
+            phone_number: phoneNumber,
+            family_name: familyName,
+            given_name: givenName,
+        } = response.data
+
+        if (response.status !== 200 || isNil(sub)) {
+            throw new Error(JSON.stringify(response.data))
+        }
+
+        return {
+            id: sub,
+            issuer: iss,
+            email: this.cleanString(email),
+            phoneNumber: this.cleanPhone(phoneNumber),
+            familyName: this.cleanName(familyName),
+            givenName: this.cleanName(givenName),
+        }
+    }
+
+    async getUserExternalIdentityId ({ idToken }) {
+        const decodedToken = jwtDecode(idToken)
+        return decodedToken.sub
+    }
+
+    generateRequestId () {
+        return nanoid()
+    }
+
+    cleanString (value) {
+        return !isNil(value) ? value.toLowerCase() : null
+    }
+
+    cleanPhone (value) {
+        return !isNil(value)
+            ? value.replaceAll(' ', '').replaceAll('(', '').replaceAll(')', '')
+            : null
+    }
+
+    cleanName (value) {
+        if (isNil(value)) {
+            return value
+        } else {
+            return value.split(' ').map(part => {
```

Comment:
replace(/\s+/g, ' ') - to exclude double spaces

Reply:
+1

Reply:
Ok